### PR TITLE
PLT-5380 Moved link preview image to top right corner of preview area

### PR DESCRIPTION
**Pull request by debanshuk**
#### Summary
Moved link preview image to top right corner of preview area

Additionally:
* Reduced max size of link preview image to 100X100 pixels (from 150X150 pixels)
* Added logic to hide image area if image loading fails

#### Ticket Link
JIRA Ticket: https://mattermost.atlassian.net/browse/PLT-5380

#### Checklist
- [x] Has UI changes

### CODE REVIEW

**Comment by jasonblais** [comment]
Thanks @debanshuk for the quick fix!

Looks like the test server isn't generating, so I'll take a look at the PR once it's resolved.

Just checking if you moved the link preview image to the top RIGHT corner, not the left?

**Comment by debanshuk** [comment]
@jasonblais, Yes I moved the it to top right corner (not left) 😶. Fixed commit message and PR description.

**Comment by jasonblais** [comment]
Hey @debanshuk, thanks again for the pull request! Just chatting with @asaadmahmood about the UI and whether any changes need to be made.

**Comment by jasonblais** [comment]
Thanks @debanshuk! Overall it looks great, there's just a few UI tweaks that would be nice to have, in order to properly accommodate moving images to the top right corner.

Sorry for not including these in the original ticket, something that I overlooked when creating it.

I'll use sample link previews as a basis for describing the improvements

1) 
image

a. Propose reducing the max height for images to 80px on desktop view; 70px on tablet/mobile view
b. Propose letting the website title wrap to more than one line (up to 300 characters), instead of truncating it as done now
c. The whitespace to the left and right of the image isn't equal. Propose increasing whitespace on the LEFT of the image to be at least a minimum of what's included on the right.

2. 
image

a. The image on the top right corner is quite small -- is there a way to increase its height to match 1a (80px; 70px in tablet/mobile view).
b. For images in landscape (horizontally longer than vertically) that have a width of at least 400px, it's okay to keep the image below the text.

3. 
image

For previews such as shown above, the container width is a little too wide resulting in a lot of whitespace between the description and the image. Previously the image was below the text, so that wasn't an issue.

To accommodate for this, can you decrease the container width for link previews to 800px? That way the link preview box won't be as wide.

**Comment by debanshuk** [comment]
@jasonblais, Changes stated in point 1 and 2 are simple enough, I'll do them. For the  subpoints in point number 2, following are my responses:

2.a: We can show image with fixed high and width but that would stretch the image in cases like this. Or we can show the image vertically centered (with a different background colour maybe), would that work?

2.b: We can do this. OG image task can have image's height and width, but they don't always have it. So, to get an image's height and width we would first have to load that image and then only we would be able to decided it's potion in the link preview and render the preview. 

i. Ie, preview generation would be delayed until the image is loaded. If that's fine, I'll do this change?
ii. Also, if the image is shown below the text would it be collapsible?
iii. And, should we show the image with it's full width (but not more than the width of preview area) when it is shown below the text area?

**Comment by jasonblais** [comment]
Sorry for my late reply:

2a:

> We can show image with fixed high and width but that would stretch the image in cases like this

I see, I thought we were scaling the image down so it fits in the top right corner.

> Or we can show the image vertically centered (with a different background colour maybe), would that work?

How do you mean?

2b: 

> i. If the image is shown below the text would it be collapsible?

Yes.

> ii. And, should we show the image with it's full width (but not more than the width of preview area) when it is shown below the text area?

How would the current implementation handle this?

**Comment by debanshuk** [comment]
@jasonblais, Following are my responses:

2a:
>>We can show image with fixed high and width but that would stretch the image in cases like this

> I see, I thought we were scaling the image down so it fits in the top right corner.
We are but image might get stretched if it's dimension doesn't match the dimension with which we are showing it as showing in following mockup
screen shot 2017-02-03 at 9 43 38 pm


 >>   Or we can show the image vertically centered (with a different background colour maybe), would that work?

> How do you mean?

Like as follows:

- Without any background colour:
screen shot 2017-02-03 at 9 46 31 pm

- With a background colour:
screen shot 2017-02-03 at 10 00 35 pm

2b: 
>>    i. If the image is shown below the text would it be collapsible?

> Yes. (edit: sorry for this additional case that was added after the fact and wasn't part of the original ticket. If you'd prefer to have 2b applied in a separate ticket, I'm totally okay with that)

No, I can do this in the same ticket. I just need to reinstate old functionality.


>>    ii. And, should we show the image with it's full width (but not more than the width of preview area) when it is shown below the text area?

> How would the current implementation handle this?

In the current implementation it would look like following screenshot:
screen shot 2017-02-03 at 10 20 24 pm


If we allow the image to get to it's width, it would look like following screenshot:
screen shot 2017-02-03 at 10 19 55 pm

**Comment by jasonblais** [comment]
2a: Got it -- what I actually mean is increasing the height to match 1a (80px; 70px in tablet/mobile view) AND keeping the aspect ratio. 

So you would increase the dimensions of the image (while keeping aspect ratio) until height matched 1a.

And let's go without the background colour.

2b: Let's keep the current implementation:
   - saves vertical space
   - allows us to test this in different scenarios without having you do code changes before then

**Comment by debanshuk** [comment]
@jasonblais, Thanks for the replies. I am doing these changes. One more small question came to my mind during develop though:

Following are your replies about showing image below the text.

> For images in landscape (horizontally longer than vertically) that have a width of at least 400px, it's okay to keep the image below the text.
>
> 2b: Let's keep the current implementation:
>
>    saves vertical space
    allows us to test this in different scenarios without having you do code changes before then

1. As per above image an image is horizontally longer than vertically and have a width of at least 400px, we should show in below the text with a max width if 150px, as in the current implementation (of showing image below the text) we only allow the image to expand upto 150 px. With the 150px restriction, do we still want the image to be larger 400px? Shouldn't we show the image below the text if it's horizontally longer than vertically and is greater than equal to 150px?

2. OpenGraph metadata of page can have multiple images with different sizes (and content). Currently the code using using nearest point algorithm to get the image which is closest in size to our desired size (earlier it was 150x150px, now it's 80x80px), with no preference to vertically or horizontally longer images. So, if a page has two images in it's OpenGraph metadata, one with size 50x70px and other with size 1500x800px, current implementation would choose the 50x70px image. Is that fine, or should we give preference to images larger than 80x80px size, but still nearest to that size? Eg, previous case use the 1500x800px image, but if three images are of size 79x79px, 90x90px and 400x200px, use 90x90px image as that is closest to 80x80px size in the images which are greater than or equal to 80x80px.

**Comment by debanshuk** [comment]
@jasonblais, I have done all the changes as per above conversation. Please review them.
Also, let me know your responses for my last comment and if I need to do any changes regarding them.

**Comment by jasonblais** [comment]
Thanks @debanshuk! I'll spin up another testing server for your changes.

I'll think about your questions a little more while I'm testing so I get a feel for how the link previews work now in different situations.

**Comment by jasonblais** [comment]
Thanks @debanshuk!  Just a couple of issues found during initial testing:

1) Looks like the images appear in the top right corner first, then with text, instead of having the text and the image side-by-side.

Sample link:  https://github.com/mattermost/platform/pull/5212
image

2) Sometimes an image is shown at the bottom, even though it's not landscape.

We might need an additional criteria to define what is "landscape". If the width-to-height ratio is 16:9 (or greater), then the image is landscape and should appear below the message. Otherwise the image should be at the top right corner.

Sample link: https://twitter.com/ArchieComics/status/813007703861841920
image

Your questions:

> Shouldn't we show the image below the text if it's horizontally longer than vertically and is greater than larger 150px horizontally?

Yes, that makes sense.

> Is that fine, or should we give preference to images larger than 80x80px size, but still nearest to that size?

Let's go with whatever image is the closest one. If we have an image with dimensions 1500x800px, then try to shrink it to 80x80px, it probably wouldn't look good. So it would be more ideal to choose the smaller image (50x70px).

**Comment by debanshuk** [comment]
@jasonblais, I have done changes/fixes as per your last comment. Please review.

**Comment by jasonblais** [comment]
Awesome! I tested it out and queued to our [next UX meeting for approval](https://pre-release.mattermost.com/core/pl/deq1f3jd9b8qtr5sfgs4dmb1da).

Great work :tada:

**Comment by jasonblais** [comment]
Hey @debanshuk all looks great! UX team approved and really likes the new changes!

There's just one minor UI defect that I didn't notice before. The whitespace at the top and and bottom of the preview isn't equal (see screenshot). Would you like to help reduce whitespace from the bottom, so that they're equal? I'll then queue it for dev review! 👍 

image
*^Observed*

image

*^Expected*

**Comment by debanshuk** [comment]
@jasonblais, I have done few changes to fix the white space issue. Please review.

**Comment by jasonblais** [comment]
Thanks @debanshuk! The change looks good.

I unfortunately discovered one new issue on mobile view. If I post the following link on mobile (or narrow desktop browser window), the preview breaks: https://github.com/mattermost/platform/issues/1899

image

Can you help take a look? Sorry I didn't notice this before, it looks like it only happens for a couple of links.

**Comment by debanshuk** [comment]
@jasonblais, The issue was occurring for links which had a long unbreakable word in description of title (eg, a link). I wasn't able to find any way to fix the issue completely using only CSS because in our currently layout width of both elements (image and text) are variable. So I had to use some JavaScript to solve the issue. Please validate.

**Comment by jasonblais** [comment]
Thanks @debanshuk! Looks good, tested on both mobile and desktop. Text wrapping also looks good 👍 

Queuing next for dev review.

@asaadmahmood can you help with CSS review?

File: config/config.json [hunk 1]
```diff
@@ -69,7 +69,7 @@
         "MaxIdleConns": 20,
         "MaxOpenConns": 300,
         "Trace": false,
-        "AtRestEncryptKey": ""
+        "AtRestEncryptKey": "awu1akbsjjs1fpso4w11x1dsbe6do4ro"
```
Comment (enahum) [inline config/config.json]: @debanshuk can you exclude the changes in this file for this PR please

Comment (debanshuk) [inline config/config.json]: @enahum, Yup. Looks like this file got commit by mistake.

**Comment by debanshuk** [comment]
@enahum, I have removed changes of config.json from the PR. Please re-review.

**Comment by asaadmahmood** [comment]
@debanshuk The images at the bottom are constrained with a max-height and max-width of 150px, if the images are going to appear so small, them I'm not sure what the reason would be to move them to the bottom, we need to increase that.

The max-width should be 400px;
And max-height should be 200px;
For the image at the bottom case.

screen shot 2017-02-25 at 6 54 06 am

**Comment by asaadmahmood** [comment]
I've done the css changes, so you don't need to do them, if you can please resolve the merge conflicts.

**Comment by jasonblais** [comment]
@debanshuk please help resolve merge conflicts and we'll have one more dev review on this?

**Pull request merged**